{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4c42a0d2",
   "metadata": {},
   "source": [
    "This file further process output from VEP and split variants into mapped and unmapped to uniprot.\n",
    "\n",
    "Specifically,\n",
    "\n",
    "With the output from VEP, do:\n",
    "1. filter the output file and retain the rows where the SWISSPROT and TREMBL columns are not be both empty. \n",
    "2. There should be only one identifier in SWISSPROT and might be multiple identifiers in TREMBL (if so separated by \",\"). \n",
    "3. If the SWISSPROT or TREMBL identifiers are in the format of XXXXXX.XXX(for example P43489.17), only retain the content before \".\" (P43489). \n",
    "4. Then choose the best UniProt among SWISSPROT and TREMBL identifiers for each row, prioritizing SWISSPROT over TREMBL, and choosing the longest TREMBL identifier if there is no SWISSPROT. \n",
    "5. Naming the column containing the best UniProt as UNIPROT. Rename the Protein_position column as UNIPROT_position.\n",
    "\n",
    "(take coding_indels as an example:)\n",
    "\n",
    "Parsing the Amino_acids column in the VEP output file into two columns: REF_AA and ALT_AA. For example, if the value in Amino_acids is \"RLTQTV/X\",\n",
    "the values in REF_AA and ALT_AA should be \"RLTQTV\" and \"X\" respectively.\n",
    "\n",
    "Retain the six columns in VEP output: #Uploaded_variation, UNIPROT, UNIPROT_position, REF_AA, ALT_AA and Consequence, and map the VEP output to\n",
    "the coding indels using #Uploaded_variation as key. \n",
    "\n",
    "Finally you should get a text file named mapped_coding_indel.txt with 18 columns: 'CHROM', 'POS', 'ID'(HGMD_ID), 'REF'(REF_N), 'ALT'(ALT_N),'REF_AA', 'ALT_AA','UNIPROT', 'UNIPROT_position','CLASS'(Variant_class),'GENE','STRAND', 'DB'(dbSNP), 'PHEN'(Phenotype),'RANKSCORE','DNA'(RefSeq_DNA_record), 'PROT'(RefSeq_protein_record), 'Consequence'\n",
    "\n",
    "Also you should get unmapped_coding_indel.txt with 13 columns (missing REF_AA, ALT_AA,\n",
    "UNIPROT, UNIPROT_position, and Consequence)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "916ae484",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "coding_indels = pd.read_csv(\"/share/yu/mw2243/HGMD/coding_indels.tsv\",sep=\"\\t\")\n",
    "VEP_res = pd.read_csv(\"/share/yu/mw2243/HGMD/indels_VEP_results.txt\",sep='\\t')\n",
    "#filter out both '-' in 'SWISSPROT','TREMBL'\n",
    "filter = VEP_res.loc[~((VEP_res['SWISSPROT'].values=='-')&(VEP_res['TREMBL'].values=='-'))]\n",
    "# Step 1: Remove version numbers (everything after '.')\n",
    "# Step 2: Select best UniProt ID with priority rules\n",
    "filter['SWISSPROT'] = filter['SWISSPROT'].str.split('.').str[0]\n",
    "filter['TREMBL'] = filter['TREMBL'].str.split('.').str[0]\n",
    "filter['UNIPROT'] = np.where(\n",
    "    filter['SWISSPROT']!='-',  # Priority 1: SWISSPROT exists\n",
    "    filter['SWISSPROT'],\n",
    "    filter['TREMBL']\n",
    ")\n",
    "# Step 3: Rename Protein_position column\n",
    "if 'Protein_position' in filter.columns:\n",
    "    filter = filter.rename(columns={'Protein_position': 'UNIPROT_position'})\n",
    "\n",
    "\n",
    "filter[['REF_AA', 'ALT_AA']] = filter['Amino_acids'].str.split('/', expand=True)\n",
    "filter = filter[['UNIPROT','UNIPROT_position','REF_AA', 'ALT_AA','Consequence']]\n",
    "df_joined = pd.concat([filter.reset_index(drop=True), coding_indels.iloc[filter.index.values,:].reset_index(drop=True)],axis=1)\n",
    "df_joined = df_joined[['CHROM', 'POS', 'ID', 'REF', 'ALT','REF_AA', 'ALT_AA','UNIPROT', 'UNIPROT_position','CLASS','GENE','STRAND', 'DB', 'PHEN','RANKSCORE','DNA', 'PROT', 'Consequence']]\n",
    "df_joined.to_csv('/share/yu/mw2243/HGMD/processed/mapped_coding_indel.txt')\n",
    "\n",
    "mask = ~coding_indels.reset_index(drop=True).index.isin(filter.index)\n",
    "filtered_indels = coding_indels.reset_index(drop=True)[mask].reset_index(drop=True)\n",
    "filtered_indels.to_csv(\"/share/yu/mw2243/HGMD/processed/unmapped_coding_indel.txt\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ec6501f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#missense\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "VEP_res_path = \"/share/yu/mw2243/HGMD/missense_VEP_results.txt\"\n",
    "missense = pd.read_csv(\"/share/yu/mw2243/HGMD/missense_variants.tsv\",sep=\"\\t\")\n",
    "VEP_res = pd.read_csv(VEP_res_path,sep='\\t')\n",
    "#filter out both '-' in 'SWISSPROT','TREMBL'\n",
    "filter = VEP_res.loc[~((VEP_res['SWISSPROT'].values=='-')&(VEP_res['TREMBL'].values=='-'))]\n",
    "filter['SWISSPROT'] = filter['SWISSPROT'].str.split('.').str[0]\n",
    "filter['TREMBL'] = filter['TREMBL'].str.split('.').str[0]\n",
    "filter['UNIPROT'] = np.where(\n",
    "    filter['SWISSPROT']!='-',  # Priority 1: SWISSPROT exists\n",
    "    filter['SWISSPROT'],\n",
    "    filter['TREMBL']\n",
    ")\n",
    "if 'Protein_position' in filter.columns:\n",
    "    filter = filter.rename(columns={'Protein_position': 'UNIPROT_position'})\n",
    "\n",
    "filter[['REF_AA', 'ALT_AA']] = filter['Amino_acids'].str.split('/', expand=True)\n",
    "filter = filter[['UNIPROT','UNIPROT_position','REF_AA', 'ALT_AA','Consequence']]\n",
    "df_joined = pd.concat([filter.reset_index(drop=True), missense.iloc[filter.index.values,:].reset_index(drop=True)],axis=1)\n",
    "df_joined = df_joined[['CHROM', 'POS', 'ID', 'REF', 'ALT','REF_AA', 'ALT_AA','UNIPROT', 'UNIPROT_position','CLASS','GENE','STRAND', 'DB', 'PHEN','RANKSCORE','DNA', 'PROT', 'Consequence']]\n",
    "df_joined.to_csv('/share/yu/mw2243/HGMD/processed/mapped_missense.txt')\n",
    "\n",
    "mask = ~missense.reset_index(drop=True).index.isin(filter.index)\n",
    "filtered_indels = missense.reset_index(drop=True)[mask].reset_index(drop=True)\n",
    "filtered_indels.to_csv(\"/share/yu/mw2243/HGMD/processed/unmapped_missense.txt\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c9c1a14",
   "metadata": {},
   "outputs": [],
   "source": [
    "#nonsense\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "VEP_res_path = \"/share/yu/mw2243/HGMD/nonsense_VEP_results.txt\"\n",
    "nonsense = pd.read_csv(\"/share/yu/mw2243/HGMD/nonsense_variants.tsv\",sep=\"\\t\")\n",
    "VEP_res = pd.read_csv(VEP_res_path,sep='\\t')\n",
    "#filter out both '-' in 'SWISSPROT','TREMBL'\n",
    "filter = VEP_res.loc[~((VEP_res['SWISSPROT'].values=='-')&(VEP_res['TREMBL'].values=='-'))]\n",
    "filter['SWISSPROT'] = filter['SWISSPROT'].str.split('.').str[0]\n",
    "filter['TREMBL'] = filter['TREMBL'].str.split('.').str[0]\n",
    "filter['UNIPROT'] = np.where(\n",
    "    filter['SWISSPROT']!='-',  # Priority 1: SWISSPROT exists\n",
    "    filter['SWISSPROT'],\n",
    "    filter['TREMBL']\n",
    ")\n",
    "if 'Protein_position' in filter.columns:\n",
    "    filter = filter.rename(columns={'Protein_position': 'UNIPROT_position'})\n",
    "\n",
    "\n",
    "filter[['REF_AA', 'ALT_AA']] = filter['Amino_acids'].str.split('/', expand=True)\n",
    "filter = filter[['UNIPROT','UNIPROT_position','REF_AA', 'ALT_AA','Consequence']]\n",
    "df_joined = pd.concat([filter.reset_index(drop=True), nonsense.iloc[filter.index.values,:].reset_index(drop=True)],axis=1)\n",
    "df_joined = df_joined[['CHROM', 'POS', 'ID', 'REF', 'ALT','REF_AA', 'ALT_AA','UNIPROT', 'UNIPROT_position','CLASS','GENE','STRAND', 'DB', 'PHEN','RANKSCORE','DNA', 'PROT', 'Consequence']]\n",
    "df_joined.to_csv('/share/yu/mw2243/HGMD/processed/mapped_nonsense.txt')\n",
    "\n",
    "mask = ~nonsense.reset_index(drop=True).index.isin(filter.index)\n",
    "filtered_indels = nonsense.reset_index(drop=True)[mask].reset_index(drop=True)\n",
    "filtered_indels.to_csv(\"/share/yu/mw2243/HGMD/processed/unmapped_nonsense.txt\")\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
